File: scripts/dump-legal-moves.ts
// Used by scripts/compare-engines.py to dump legal moves for a fixture
// position as JSON, in a format directly comparable to the Python engine's
// dumper output.
//
// Usage: npx tsx scripts/dump-legal-moves.ts <fixture.json> <out.json>

import { readFileSync, writeFileSync } from "node:fs";
import type { Board } from "../src/engine/board";
import { legalMoves, type Move } from "../src/engine/moves";

const [, , fixturePath, outPath] = process.argv;
if (!fixturePath || !outPath) {
  console.error("usage: dump-legal-moves.ts <fixture.json> <out.json>");
  process.exit(1);
}

function sortMoves(moves: Move[]): string[] {
  return moves
    .map((move: Move) => `${move.from.rank},${move.from.file}->${move.to.rank},${move.to.file}`)
    .sort();
}

const { board } = JSON.parse(readFileSync(fixturePath, "utf8")) as { board: Board };

const result: Record<"w" | "b", string[]> = {
  w: sortMoves(legalMoves(board, "w")),
  b: sortMoves(legalMoves(board, "b")),
};

writeFileSync(outPath, JSON.stringify(result, null, 2));

File: python-engine/dump_legal_moves.py
"""CLI used by ../scripts/compare-engines.py to dump legal moves for a fixture
position as JSON, in a format directly comparable to the TS dumper's output.

Usage: python dump_legal_moves.py <fixture.json> <out.json>
"""

import json
import sys

from board import Board, Piece
from moves import Move, legal_moves


def load_board(path: str) -> Board:
    with open(path, "r", encoding="utf-8") as f:
        data = json.load(f)

    board: Board = []
    for row in data["board"]:
        board.append([None if cell is None else Piece(cell["color"], cell["type"]) for cell in row])
    return board


def sort_moves(moves: list[Move]) -> list[str]:
    return sorted(f"{m.frm.rank},{m.frm.file}->{m.to.rank},{m.to.file}" for m in moves)


def main() -> None:
    fixture_path, out_path = sys.argv[1], sys.argv[2]
    board = load_board(fixture_path)

    result = {
        "w": sort_moves(legal_moves(board, "w")),
        "b": sort_moves(legal_moves(board, "b")),
    }

    with open(out_path, "w", encoding="utf-8") as f:
        json.dump(result, f, indent=2)


if __name__ == "__main__":
    main()
